tchannel/tornado/peer.py

```diff
@@ -735,14 +735,19 @@ def choose(self, hostport=None, blacklist=None):
         if hostport:
             return self.get(hostport)
 
-        hosts = self._peers.viewkeys() - blacklist
-        if not hosts:
-            return None
+        pop_peers = deque()
+        choose_peer = None
+        n = self.peer_heap.size()
 
-        peers = list(self._connected_peers(hosts))
-        if peers:
-            return peers[random.randint(0, len(peers)-1)]
-        else:
-            hosts = list(hosts)
-            host = hosts[random.randint(0, len(hosts)-1)]
-            return self.get(host)
+        while n > 0:
+            peer = self.peer_heap.pop_peer()
+            pop_peers.append(peer)
```

Comment:
I implemented this in #345 for merging into the this branch. You should be able to use it here.